### PR TITLE
Add fuzzy search hook to `@automattic/search`

### PR DESCRIPTION
**Pull request by zaguiini**
#### Proposed Changes

Right now, our client-side searches are pretty limited: we have the `searchSites` HOC that exposes basic search functionality that's not very customizable and the results have to match the term exactly, bringing a subpar user experience.

This PR adds a `useFuzzySearch` hook that relies on the `fuse.js` package. That hook can be used anywhere search is needed. You can pass fields that are indexed and the search state is also handled.

#### Testing Instructions

Check out this branch, run `yarn install` then `yarn build-packages` so the `@automattic/search` package gets updated.

Once done, apply this diff to Calypso:

```diff
diff --git a/client/components/search-sites/index.js b/client/components/search-sites/index.js
index e17a05af82..f1fa99da83 100644
--- a/client/components/search-sites/index.js
+++ b/client/components/search-sites/index.js
@@ -1,3 +1,4 @@
+import { useFuzzySearch } from '@automattic/search';
 import { get } from 'lodash';
 import { Component } from 'react';
 import { connect } from 'react-redux';
@@ -16,27 +17,20 @@ const mapState = ( state ) => ( {
 export default function searchSites( WrappedComponent ) {
 	const componentName = WrappedComponent.displayName || WrappedComponent.name || '';
 
-	class Searcher extends Component {
-		state = { term: null };
-
-		setSearchTerm = ( term ) => this.setState( { term } );
-
-		getSearchResults() {
-			return this.state.term
-				? searchCollection( this.props.sites, this.state.term.toLowerCase(), [ 'name', 'URL' ] )
-				: null;
-		}
-
-		render() {
-			return (
-				<WrappedComponent
-					{ ...this.props }
-					searchSites={ this.setSearchTerm }
-					sitesFound={ this.getSearchResults() }
-				/>
-			);
-		}
-	}
+	const Searcher = ( props ) => {
+		const { query, setQuery, results } = useFuzzySearch( {
+			data: props.sites,
+			fields: [ 'name', 'URL' ],
+		} );
+
+		return (
+			<WrappedComponent
+				{ ...props }
+				searchSites={ setQuery }
+				sitesFound={ query ? results : null }
+			/>
+		);
+	};
 
 	const Connected = connect( mapState )( Searcher );
 	Connected.displayName = `SearchSites(${ componentName })`;
```

Open any site in Calypso, click `Switch Site` and start typing in the search bar. It no longer requires an exact match to return queried results.

This is an optional change to the `searchSites` HOC just to exemplify the usage. It does not necessarily need to be used there. I first implemented this hook in #65454 but I moved it here so I don't have a PR breaking the single-subject rule.

Related to #65169, first implemented in #65454.

### CODE REVIEW

**Comment by danielbachhuber** [comment]
> This PR adds a `useFuzzySearch` hook that relies on the `fuse.js` package. That hook can be used anywhere search is needed. You can pass fields that are indexed and the search state is also handled.

@zaguiini What are some examples where fuzzy search (and a separate library) are advantageous over the pure string match in #65505 ?

**Comment by vykes-mac** [comment]
I'm not sure what's performance impact but I think fuzzy can be beneficial especially as list of sites grow. I might want to find a site that i do not remember the exact name but entering partial string finds the site for me.

Is the search on cached results or any server call is made?

**Comment by vykes-mac** [comment]
I actually like this feature, there are particular sites I use at times but after I while it gets lost in recently used order, I often don't remember the name especially with the generated free domain names. Testing this out made it easy to find a site quickly that would otherwise take me sometime.

**Comment by zaguiini** [comment]
Thanks, @vykes-mac. That's exactly the case I think fuzzy search should be used for.

Furthermore, there's this great article [here](https://www.algolia.com/blog/engineering/discover-what-fuzzy-search-is-with-fuzzy-matching/) exemplifying use cases and why it's a good idea.

So @danielbachhuber, the difference is mostly subjective as there are no performance benefits of one over the other.

File: package.json
Comment (danielbachhuber) [inline package.json]: Do we need to follow any special evaluation, etc. process when adding a new dependency to Calypso?

**Review by danielbachhuber — APPROVED** [review]
Works great!

image

I don't feel I have any qualified opinions on the JavaScript fanciness, so it'd be good to get a second pair of eyes in that regard.

Also worth noting: I had to add a newline in order for the patch to apply.

File: package.json
Comment (p-jackson) [reply package.json]: I also think we should add it to `packages/search/package.json`

> You should add dependencies to the root project _only_ when it will be used to test and/or build other packages.

https://github.com/Automattic/wp-calypso/blob/trunk/docs/dependency-management.md

Comment (tyxla) [reply package.json]: We actually should add it to `packages/search` and remove it from here if you ask me.

File: packages/search/src/use-fuzzy-search.ts
Comment (tyxla) [inline packages/search/src/use-fuzzy-search.ts:23]: This hook could use some tests!

**Comment by zaguiini** [comment]
@tyxla: added some tests.

One enhancement is now that we're reusing the same `Fuse` instance whenever `data` changes, instead of throwing everything away. That improves performance A LOT: a **6x CPU slowdown** allows removing debouncing on the `<Search />` instance and it still queries very quickly. I'm honestly impressed!

Try it yourself:

```diff
diff --git a/client/components/site-selector/index.jsx b/client/components/site-selector/index.jsx
index e010714eb9..a085c03c9e 100644
--- a/client/components/site-selector/index.jsx
+++ b/client/components/site-selector/index.jsx
@@ -416,7 +416,6 @@ export class SiteSelector extends Component {
 			>
 				<SearchComponent
 					onSearch={ this.onSearch }
-					delaySearch={ true }
 					placeholder={ this.props.searchPlaceholder }
 					// eslint-disable-next-line jsx-a11y/no-autofocus
 					autoFocus={ this.props.autoFocus }
```

Then open the sidebar > site search in Calypso.

Please, re-review the PR!

File: packages/search/src/index.ts
Comment (danielbachhuber) [inline packages/search/src/index.ts:2]: Should we document usage in `packages/search/README.md`?

I think it'd be really nice to have some examples of normal search vs. fuzzy search.

File: packages/search/src/test/use-fuzzy-search.tsx
Comment (danielbachhuber) [inline packages/search/src/test/use-fuzzy-search.tsx:7]: Out of curiosity (I'm new to TypeScript), what does `< T, >` designate?

File: packages/search/src/use-fuzzy-search.ts
Comment (danielbachhuber) [inline packages/search/src/use-fuzzy-search.ts:19]: Out of curiosity (I'm new to TypeScript), what does `< Fuse.IFuseOptions< T > >` designate?

**Review by danielbachhuber — APPROVED** [review]
Works well in my testing! Left some non-blocking comments.

File: packages/search/src/test/use-fuzzy-search.tsx
Comment (p-jackson) [reply packages/search/src/test/use-fuzzy-search.tsx:7]: I just learnt this trick yesterday myself! The short story is that it means the same thing as `< T >` because the trailing slash will be ignored. So it's signifying that this anonymous function is [generic](https://www.typescriptlang.org/docs/handbook/2/generics.html).

The reason for the comma is that it helps work around a slight incompatibility between the TypeScript and JSX languages. When the JSX parser sees `< T >` it thinks it's a React element. When the TypeScript parser sees `< T >` it thinks it signifies some sort generic. Usually it's possible to disambiguate between the two:

```
// These couldn't possibly be React elements because of their context
function foo< A >....
const m = new Set< number >...

// But what about this one ....
const a = < B >...
```

The rule is that the last example will always be a React element and will never represent something like a generic anonymous function (in `.tsx` files at least, I think the rules are different in `.ts` files). You can disambiguate for the parser by doing something like `< B, >` or `< B extends Sometype >`. Now it knows it must be TypeScript syntax.

Anyway, that's why in `.tsx` files I think it's a lot more common to see non-anonymous `function` functions. There are fewer weird rules. It's also why the TypeScript compiler _insists_ that JSX can only be written in `.tsx` files, unlike Babel which will accept JSX syntax in `.js` files.

File: packages/search/src/use-fuzzy-search.ts
Comment (p-jackson) [reply packages/search/src/use-fuzzy-search.ts:19]: `IFuseOptions< T >` is the type the `Fuse` class constructor takes as a second argument. https://github.com/krisk/Fuse/blob/7ff9c2f2ea8345287b05f7560799c6fee92325e3/src/index.d.ts#L7

`T` here represents the data type that we're searching. Since the Fuse library supports searching arbitrary data types, it means we have to refer to it as an arbitrary type (a.k.a. a "generic" type). `IFuseOptions` needs to know what that generic type is though because it changes what the options object looks like, that's why we pass the `T` type to it like `IFuseOptions< T >`.

Sometimes `T` is just `string`, like in the first unit test. And we're just searching through an array of strings. Sometimes (like in the dashboard), `T` will be an object like `{ title: string, domain: string, ... }` since we'll be searching through the array of site data objects. Traditionally generics use short identifiers like `T` or `U` for some reason, but they can be longer. Maybe something like `Needle` would be better? `data: Needle[]` is the entire haystack we're searching, and this hook returns `results: Needle[]` is all of the needles we found using this particular query.

The final thing is `Partial< ... >` which is a built-in TypeScript helper. It makes all of the `IFuseOptions` optional. It probably isn't necessary because all the optional already are optional https://github.com/krisk/Fuse/blob/7ff9c2f2ea8345287b05f7560799c6fee92325e3/src/index.d.ts#L242

File: packages/search/src/index.ts
Comment (p-jackson) [reply packages/search/src/index.ts:2]: Pasting a couple of the unit test examples into the README sounds good.

File: packages/search/src/use-fuzzy-search.ts
Comment (p-jackson) [inline packages/search/src/use-fuzzy-search.ts]: The explicit generic parameters aren't needed here and it might be better to leave them out. It allows the libraries to change how their type definitions work, and as long as they continue to support type inference, our code will just keep working.

```suggestion
	const [ fuseInstance ] = useState( () => {
		return new Fuse( data, {
```

File: packages/search/tsconfig.json
Comment (p-jackson) [inline packages/search/tsconfig.json]: I'm getting type errors in the tests file for missing definitions for `describe` and `it`, which I'm guessing is why you excluded the tests here?

I don't fully understand why some Jest files seem to get the `describe` definitions automatically and some don't, but I tried fixing this locally by adding `types: [ "jest" ]` to the `compilerOptions` and it seemed to do the trick.

**Review by p-jackson — APPROVED** [review]
Looks like Fuse supports some powerful searching. Maybe in the future we'll want to expose the scores from the hook too, but we can do that when the need arises.

`fuse.js` [seems solid enough from a maintenance point of view](https://snyk.io/advisor/npm-package/fuse.js). Apache licensed.

File: packages/search/src/test/use-fuzzy-search.tsx
Comment (tyxla) [inline packages/search/src/test/use-fuzzy-search.tsx]: Ideally, in each test, we should first use `userEvent.setup()` first and then use the event triggering helpers provided by it. See p4TIVU-ac2-p2 for more details - scroll down to the "User actions" section.

This comment applies to several instances in this file.

File: packages/search/src/test/use-fuzzy-search.tsx
Comment (tyxla) [inline packages/search/src/test/use-fuzzy-search.tsx]: Generally, we try to be a bit more specific than `toBeTruthy()` and use `.toBeVisible()` or `.not.toBeInTheDocument()`.  See p4TIVU-ac2-p2 for more details - scroll down to the "jest-dom matchers" section.

This comment applies to several instances in this file.

File: packages/search/src/test/use-fuzzy-search.tsx
Comment (tyxla) [inline packages/search/src/test/use-fuzzy-search.tsx]: Do we need the `React` import for anything? I thought we got rid of those at some point.

File: packages/search/src/use-fuzzy-search.ts
Comment (tyxla) [inline packages/search/src/use-fuzzy-search.ts:61]: Do we need any sort of sanitization of the data before we pass the query to fuse?

**Review by tyxla — APPROVED** [review]
Looking good 👍 

Left a few comments, nothing blocking, and this can be shipped with or without addressing them IMHO.

Nice work 🚀

File: packages/search/src/use-fuzzy-search.ts
Comment (zaguiini) [reply packages/search/src/use-fuzzy-search.ts:61]: I don't think so, no. We're passing a string and Fuse does its checks before processing. We should be safe there.

File: packages/search/src/index.ts
Comment (zaguiini) [reply packages/search/src/index.ts:2]: Looking at the `README`, this package is responsible for providing the `<Search />` component and that's it. We have two ways of going here:

1. Modify the `README` so that it's clear that not only the component is exported, but also a hook that allows fuzzy searching. I thought about changing from "A search input component" to "Search utilities" and then include new headings showcasing the component and the hook;
2. Create a new package that only exposes the fuzzy search hook. This way we'd isolate things a bit and the README would be focused on explaining just the fuzzy search functionality.

Which one do you prefer? There are cases for both sides here.

`1` is all about the hook being very closely related to the component domain-wise, so it makes sense to stick it there. However, the `<Search />` component does not perform the search itself, it's just the input. The actual search mechanism is decoupled and totally set up by the client, so it's not like the component is going to use the hook.

So in this case, `2` would actually be better since they're domain-wise related, but they don't necessarily need to be used together. But at the same time, I'm not sure about granularity. Do we want to group by domain or by functionality? How hard is it to create a new package, and does it make sense? Looking for input here.

cc @tyxla

Comment (tyxla) [reply packages/search/src/index.ts:2]: I'm fine with either, TBH. If these can (and need to) be functionally and architecturally decoupled in a proper way, they can totally be separate packages. I can see how fuzzy search is separate, but I can also see it as an attachment to the broader search package.

Comment (zaguiini) [reply packages/search/src/index.ts:2]: Yep, yep. That makes total sense. I'll go with #1.

**Comment by zaguiini** [comment]
@tyxla @danielbachhuber @vykes-mac @p-jackson would appreciate a review of the `README` changes before merging this PR!

File: packages/search/src/test/use-fuzzy-search.tsx
Comment (danielbachhuber) [reply packages/search/src/test/use-fuzzy-search.tsx:7]: 👍 Makes sense. Thanks for the explanation, @p-jackson!

File: packages/search/README.md
Comment (danielbachhuber) [inline packages/search/README.md]:
```suggestion
## What's Included?
```

File: packages/search/README.md
Comment (danielbachhuber) [inline packages/search/README.md]:
```suggestion
`useFuzzySearch` can be beneficial especially as the sites of the list of sites grow. Some users might want to find a site that they do not remember the exact name, so they can enter a partial string, or even one with typos, and the result would still be returned.
```

File: packages/search/README.md
Comment (danielbachhuber) [inline packages/search/README.md]:
```suggestion
`useFuzzySearch` can be beneficial especially as the size of the dataset grows because this mechanism does not walk through the whole list. Also, some users might want to find an object that they do not remember the exact name, so entering a partial string, or even one with typos, would still return results ranked by proximity.
```

**Review by danielbachhuber — COMMENTED** [review]
README update looks great!

File: packages/search/README.md
Comment (danielbachhuber) [reply packages/search/README.md]: This suggested change looks like it might not apply directly...

File: packages/search/package.json
Comment (tyxla) [inline packages/search/package.json:56]: Noting that such a change would generally be a good candidate for its own PR.

File: packages/search/src/test/use-fuzzy-search.tsx
Comment (tyxla) [inline packages/search/src/test/use-fuzzy-search.tsx]: Nit: When we're querying for something that the user sees, I find `.toBeVisible()` a bit better, because in practice something might be in the document, but still be invisible for the user. At the same time, `.not.toBeInTheDocument()` is just fine.

**Review by tyxla — APPROVED** [review]
Still looking great! Thanks for the thorough README and for the tests. 

Just left a couple of little suggestions, but nothing blocking.

🚀

File: packages/search/package.json
Comment (zaguiini) [reply packages/search/package.json:56]: Yeah -- the only reason it's not on its own PR is because it's only used in the tests and scoped to that package. The tests are still passing so I thought of keeping it as simple as possible.

Comment (tyxla) [reply packages/search/package.json:56]: Yup, feel free to keep it as-is, no need to further delay this PR.

**Pull request merged**
